funsor/contract.py

```diff
@@ -0,0 +1,141 @@
+from __future__ import absolute_import, division, print_function
+
+from collections import OrderedDict
+
+import opt_einsum
+
+import funsor.ops as ops
+from funsor.distributions import Gaussian, Delta
+from funsor.optimizer import Finitary, optimize
+from funsor.sum_product import _partition
+from funsor.terms import Funsor, Number, Variable, eager
+from funsor.torch import Tensor
+
+
+# TODO handle Joint as well
+ATOMS = (Tensor, Gaussian, Delta, Number, Variable)
+
+
+def _order_lhss(lhs, reduced_vars):
+    assert isinstance(lhs, Finitary)
+
+    components = _partition(lhs.operands, reduced_vars)
+    root_lhs = Finitary(ops.mul, tuple(components[0][0]))
+    if len(components) > 1:
+        remaining_lhs = Finitary(ops.mul, tuple(t for c in components[1:] for t in c[0]))
```

Comment:
Note that since these are undirected, the optimizer should be using cost to decide which term is the root